torch/_dynamo/guards.py

```diff
@@ -455,7 +455,15 @@ def combine_scopes(left, right):
         )
         global_builder = GuardBuilder(self.id_ref, f_globals, self, renames=False)
         for guard in sorted(guards or [], key=Guard.sort_key):
-            if not config.guard_nn_modules and guard.is_nn_module():
+            if (
+                not config.guard_nn_modules
+                and guard.is_nn_module()
+                # we must guard on default args for functions,
+                # even if we specialize on the rest of the nnmodule contents
+                # TODO(whc) - presumably we can use the 'DefaultsSource' in some way here instead of name parsing?
+                and "__defaults__" not in guard.name
+                and "__kwdefaults__" not in guard.name
+            ):
```

Comment:
This is not actually as offensive as I thought it would be. We could use tagging here, or a new kind of guard (then you can check those as types).

One thing I'm wondering - if these are `DefaultsSource` now, we could always detect that when we make the guard, and write something like `guard.is_default()` as an analogue to `guard.is_nn_module()` (Or even add a guardsource subtype?). 

No strong opinion, but if this read:

```
if (
                not config.guard_nn_modules
                and guard.is_nn_module()
                and guard.is_default()
```

That would be way cleaner :)

Reply:
hmm, you can't extend an enum (GuardSource) right? and i'm not sure if it makes sense to add a new enum value for defaults, since defaults can also come from a global function or a local function etc. but if it does make sense, i could add GuardSource.DEFAULTS and go that way. i'll keep this as a next PR and land the rest as is.

Reply:
Sorry, extend as in add a new value to, not via inheritance, but doing it in a later PR is fine too.

Reply:
If we leave it as is, let's add a comment referencing DefaultSource.